data/tutorials/language/0lg_00_basic_datatypes.md

```diff
@@ -0,0 +1,844 @@
+```
+
+We can write a function to multiply out expressions of the form `n * (x + y)`
+or `(x + y) * n`, and for this we will use a nested pattern:
+
+```ocaml env=expr
+# let rec distribute = function
+  | Times (e1, Plus (e2, e3)) ->
+     Plus (Times (distribute e1, distribute e2),
+           Times (distribute e1, distribute e3))
+  | Times (Plus (e1, e2), e3) ->
+     Plus (Times (distribute e1, distribute e3),
+           Times (distribute e2, distribute e3))
+  | Plus (e1, e2) -> Plus (distribute e1, distribute e2)
+  | Minus (e1, e2) -> Minus (distribute e1, distribute e2)
+  | Times (e1, e2) -> Times (distribute e1, distribute e2)
+  | Divide (e1, e2) -> Divide (distribute e1, distribute e2)
+  | Var v -> Var v;;
+val distribute : expr -> expr = <fun>
+```
+
+This is how it can be used:
+```ocaml env=expr
+# e |> distribute |> to_string |> print_endline;;
+((n * x) + (n * y))
+- : unit = ()
+```
+
+How does the `distribute` function work? The key is in the first two patterns.
+The first pattern is `Times (e1, Plus (e2, e3))` which matches expressions of
+the form `e1 * (e2 + e3)`. The right-hand side of this first pattern is the
+equivalent of `(e1 * e2) + (e1 * e3)`. The second pattern does the same thing,
+except for expressions of the form `(e1 + e2) * e3`.
+
+The remaining patterns don't change the expressions's form, but they call the
+`distribute` function recursively on their subexpressions. This ensures that all
+subexpressions within the expression get multiplied out too (if you only wanted
+to multiply out the very top level of an expression, then you could replace all
+the remaining patterns with a simple `e -> e` rule).
+
+The reverse operation, i.e., factorising out common subexpressions, can be
+implemented in a similar fashion. The following version only works for the toplevel expression.
+```ocaml env=expr
+# let top_factorise = function
+  | Plus (Times (e1, e2), Times (e3, e4)) when e1 = e3 ->
+     Times (e1, Plus (e2, e4))
+  | Plus (Times (e1, e2), Times (e3, e4)) when e2 = e4 ->
+     Times (Plus (e1, e3), e4)
+  | e -> e;;
+val top_factorise : expr -> expr = <fun>
+# top_factorise (Plus (Times (Var "n", Var "x"),
+                   Times (Var "n", Var "y")));;
+- : expr = Times (Var "n", Plus (Var "x", Var "y"))
+```
+
+The factorise function above introduces another feature: *guards* to each
+pattern. It is the conditional that follows the `when`, and it means that
+the return code is executed only if the pattern matches and the condition in the
+`when` clause is satisfied.
+
+## Conclusion
+
+This tutorial has provided a comprehensive overview of the basic data types in OCaml and their usage. We have explored the built-in types, such as integers, floats, characters, lists, tuples, and strings, and the user-defined types: records and variants. Records and tuples are mechanisms for grouping heterogeneous data into cohesive units. Variants are a mechanism for exposing heterogeneous data as coherent alternatives.
```

Comment:
Should the comments/definitions of Records and tuples be in their respective sections (both in this paragraph and the one beneath)? Is this repetitive here, or is the intention to summarise what was already said to remind the reader? If so, I'd like to reword it slightly. 

Currently it reads as if we're getting this information for the first time.